Fix scaffolding generations

stevenh opened the pull request:
## Proposed change
    
Fix `script.scaffold` generation which is not `integration` for example `config_flow` which was failing due to hassfest quality check.

Fix `config_flow_helper` which would generate out of order manifest keys due to lack of sorting.

This moves the `sort_manifest` from to `script.util` as it's now used in multiple scripts.

This updates all flows to pass the integration path to hassfest, so it only checks the integration it's working on and skips the quality check which will fail on newly integrations.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

This fixes the issue flagged by @balloob [here](https://github.com/home-assistant/core/pull/138247#issuecomment-2666743947)

I've tested that all templates for `script.scaffold` successfully generate a new integration without error.  

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

stevenh committed:
Fix scaffolding generations

Fix `script.scaffold` generation which is not `integration` for example
`config_flow` which was failing due to hassfest quality check.

Fix `config_flow_helper` which would generate out of order manifest
keys due to lack of sorting.

This moves the `sort_manifest`` from to `script.util` as it's now used
in multiple scripts.

This updates all flows to pass the integration path to hassfest, so it
only checks the integration it's working on and skips the quality check
which will fail on newly integrations.

If a sub process fails during the generation, it will now report and
instruct the user to take action.

stevenh commented:
@balloob here's the fix for the additional issue you [mentioned on my previous PR](https://github.com/home-assistant/core/pull/138247#issuecomment-2666743947)

MartinHjelmare commented on script/util.py:
```diff
@@ -13,3 +14,23 @@ def valid_integration(integration):
+
+def sort_manifest(manifest: dict[str, Any]) -> bool:
+    """Sort manifest."""
+    keys = list(manifest.keys())
```
```suggestion
    keys = list(manifest)
```

stevenh replied:
Copied from the original, thanks for the simplification

balloob commented on script/scaffold/__main__.py:
```diff
@@ -60,28 +60,24 @@ def main() -> int:
     # Always output sub commands as the output will contain useful information if a command fails.
     print("Running hassfest to pick up new information.")
-    subprocess.run(hassfest_args, check=True)
+    subprocess.run(
```
can we catch the exception so we don't print a stack trace if it were to go wrong. Maybe even print an error that this is a bug and should be reported?

stevenh replied:
Have refactored to add error checking and reporting so the output is more understandable for users.

MartinHjelmare commented:
For the future, please don't squash or amend commits after review has started to make it easier for readers to track changes.

MartinHjelmare commented on script/scaffold/__main__.py:
```diff
@@ -98,14 +130,13 @@ def main() -> int:
 
     if args.develop:
         print("Running tests")
-        print(f"$ python3 -b -m pytest -vvv tests/components/{info.domain}")
```
The full pytest command is no longer printed now. Is that intended?

stevenh replied:
I removed to make it consistent with all the others, if we think it's useful to have this output we could add `run_process` so it's consistent. Thoughts?

MartinHjelmare replied:
Yeah, I think it's nice to see exactly what commands are being run.

stevenh replied:
Done, tweaked the output while there too.

stevenh commented:
> For the future, please don't squash or amend commits after review has started to make it easier for readers to track changes.

Thanks for info, different teams work differently 😄

Do you prefer merge commits or `Update branch` from the github UI?

Any other preferences I can read up on?

MartinHjelmare commented:
Rebasing and merging, to update the branch, are both ok, but I think most people prefer merge since history isn't lost. We'll squash when merging the PR, so it doesn't matter for the result in the target branch. It's fine to press the update branch button in the GitHub web UI.

stevenh committed:
chore: improve scaffold logging

Improve the logging of the scaffold script to make it easier to debug
problems with the script, this now includes details of each subcommand
that is run.
